Integrate Unity and Cmock with project files

Abnaby opened the pull request:
### ISSUE-ID: #95 

### Summary:
Integrating the Unity and CMock frameworks into the project. The goal is to enhance unit testing capabilities by incorporating these frameworks, allowing for better isolation and testing of individual components.

### Changes :
- Added Unity framework to the project, under path `/workspace/tools/frameworks/CMock/vendor/unity`  
- Integrated CMock framework `tools/frameworks/CMock` to facilitate the generation of mock files for seamless unit testing.
configuration at `/workspace/tools/build/CMake/cmockConfig.yml`

### Dependencies:
- `apt install ruby`

### How to Test:
1. `git clone -b Integrate_Unity_CMock https://github.com/Abnaby/Dev.git`
2. `git submodule init && git submodule update --recursive`
3. `cd tools/frameworks/CMock/`
4. `git config --global --add safe.directory /workspace/tools/frameworks/CMock`
5. `git submodule init && git submodule update --recursive`
6. `cd /workspace`
7. `make unit_test rebuild=1`
### Output
![image](https://github.com/autosarzs/Dev/assets/31739953/a427b912-9d77-448e-936e-ecbbdcd5d8e3)

autosarzs commented on .github/workflows/scheduled.yml:
```diff
@@ -3,7 +3,7 @@ name: Scheduled Build
     branches:
-      - main
+      - Integrate_Unity_CMock
 
```
Don't change the branch name because this is for main branch

1- If you did that because you wanted to test CI for your branch its okay, but then you should have reverted it back
2-please rebase to main because I've created workflow for pull request, so an automated checks and tests will be running on your pull request you don't have to touch scheduled

autosarzs reviewed (commented):
Rebase to main

autosarzs commented on Makefile:
```diff
@@ -31,22 +31,37 @@ CMAKE_TEST_VARIABLES += -DCMAKE_PROJECT_DIR=$(CMAKE_PROJECT_DIR)
+		echo "Running $$program"; \
+		$$program -v | sed -E 's#.*\.c:[0-9]+:TEST\(([^)]+)\)#TEST(\1)#g; s#/workspace/software/bsw/mcal/unity_swc/tests/##g'; \
+	done
 
```
you removed the installation to CMake/output/tests 

We need test results reports uploaded to CI

Abnaby replied:
Added the installation process until it is replaced by gcovr reports.

autosarzs commented on .github/workflows/scheduled.yml:
```diff
@@ -42,20 +42,33 @@ jobs:
       - name: Checkout code
         uses: actions/checkout@v4
 
+        # Run submodule initialization and update within the devcontainer
```
action/checkout@v4 supports checkout recursive
https://stackoverflow.com/a/65091631

autosarzs commented on software/bsw/mcal/dio_swc/tools/cmake/dio_swc_tests.cmake:
```diff
@@ -2,31 +2,54 @@ set(
 set(
     dio_swc_ut_sources_list
-
+    
```
to decrease the CMake dependency fingerprint we can export one variable with common sources and headers and include it like this 

add_executable(dio_swc_test ${dio_swc_tests_sources_list} ${unity_common_sources})

Abnaby commented:
Updates 
 - Rebase
 - added installation to `CMake/output/tests` until integrating gcovr tool (will be replaced by gcovr reports)
- added unity_common_sources and unity_common_includes in the user.cmake file.
- Configured mock files to be generated under `/workspace/tools/build/CMake/Test/${module_name}/mocks`, ensuring that any cached mock files are ignored.

Dependencies:
- `apt install ruby`

mr0xc0der commented on .gitmodules:
```diff
@@ -2,3 +2,9 @@
 	path = wiki
 	url = https://github.com/autosarzs/Dev.wiki.git
 	branch = master
+[submodule "CMock"]
```
I believe that the duplication of submodule was done by mistake, could you please confirm?

Abnaby replied:
yes, it was done by mistake

mr0xc0der commented:
I would suggest to use a pre-defined folder structure for the unit tests for example tests/unit to avoid confustion in the future with other testing mechanisms (integration, e2e, ..etc)
@autosarzs @Abnaby

Abnaby commented:
Note 
--------
This pull request has been closed as the associated branch name has been changed for better alignment with our naming conventions and workflow improvements. 

For continued discussion and review, please refer to the new pull request created under the updated branch name: [#96].

Thank you!

Closed in favor of #96
